### PR TITLE
Adding section: Storage Elements

### PR DESCRIPTION
**Pull request by gusthoff**

### CODE REVIEW

File: content/courses/advanced-ada/parts/resource_management/memory_management.rst [hunk 1]
```diff
@@ -263,15 +263,91 @@ number of components.
 Storage elements
 ----------------
 
-.. admonition:: Relevant topics
+We saw parts of the :ada:`System.Storage_Elements` package while discussing
+about :ref:`addresses <Adv_Ada_Addresses>`. However, we haven't discussed yet
```
Comment (pat-rogers) [inline content/courses/advanced-ada/parts/resource_management/memory_management.rst:267]: Strike "about"

File: content/courses/advanced-ada/parts/resource_management/memory_management.rst [hunk 2]
```diff
@@ -263,15 +263,91 @@ number of components.
...
+definition:
 
-    Complete section!
+.. code-block:: ada
+
+    type Storage_Array is
+      array (Storage_Offset range <>) of
+        aliased Storage_Element;
+
+A storage array is used to represent a contiguous sequence of storage elements
```
Comment (pat-rogers) [inline content/courses/advanced-ada/parts/resource_management/memory_management.rst:285]: Maybe stress that it is contiguous? IIRC that is the only language-defined array type that has that guarantee.

File: content/courses/advanced-ada/parts/resource_management/memory_management.rst [hunk 3]
```diff
@@ -263,15 +263,91 @@ number of components.
...
+in memory. In other words, you can think of an object of :ada:`Storage_Array`
+type as a (memory) buffer.
+
+.. admonition:: For further reading
+
+    Note that the :ada:`Storage_Offset` is a range type. It's used not only
```
Comment (pat-rogers) [inline content/courses/advanced-ada/parts/resource_management/memory_management.rst:291]: Should "range type" be "integer type"?  I'm not familiar with the term "range type" anyway...

Comment (gusthoff) [reply content/courses/advanced-ada/parts/resource_management/memory_management.rst]: Yes, indeed, I've just changed it.

File: content/courses/advanced-ada/parts/resource_management/memory_management.rst (hunk 2 above)
Comment (gusthoff) [reply content/courses/advanced-ada/parts/resource_management/memory_management.rst:285]: Ok, I've just included a small admonition about it, and mentioned storage pools.

(Unfortunately, AARM 3.6 is not explicit about arrays being contiguous or not.)

**Comment by pat-rogers** [comment]
On 29-Dec-23 9:50 AM, Gustavo A. Hoffmann wrote:
> ***@***.**** commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In 
> content/courses/advanced-ada/parts/resource_management/memory_management.rst <https://github.com/AdaCore/learn/pull/971#discussion_r1438312372>:
> 
>>  
> -    Complete section!
> +.. code-block:: ada
> +
> +    type Storage_Array is
> +      array (Storage_Offset range <>) of
> +        aliased Storage_Element;
> +
> +A storage array is used to represent a contiguous sequence of storage elements
> 
> Ok, I've just included a small admonition about it, and mentioned 
> storage pools.
> 
> (Unfortunately, AARM 3.6 is not explicit about arrays being contiguous 
> or not.)

See AARM 13.7.1 paragraph 12 (the last sentence of 12).

Paragraph 12.a is also potentially useful rationale to include in your 
existing text that says that Storage_Offset is an integer type.

**Comment by gusthoff** [comment]
> See AARM 13.7.1 paragraph 12 (the last sentence of 12).
> Paragraph 12.a is also potentially useful rationale to include in your
> existing text that says that Storage_Offset is an integer type.

Yes, that's actually where I got the "a contiguous sequence of storage elements" part (in my text) from. 

However, I was now looking for any references in AARM to arrays (in general, not just `Storage_Array` type) being allocated as a contiguous block. What I understand is that the language guarantees "contiguous allocation" via `Allocate`, but this is not specific to array types.

**Review by pat-rogers — APPROVED** [review]

**Pull request merged**